concepts/errors/about.md

```diff
@@ -0,0 +1,134 @@
+
+```crystal
+raise "This is an error"
+```
+
+There are several built-in exceptions in Crystal, like `ArgumentError`, `IndexError`, `KeyError`, `IOError`, `SystemCallError`, `TypeError`, `ZeroDivisionError` and many more, these require you to pass a message to the exception.
+
+```crystal
+raise ArgumentError.new("This is an argument error")
+```
+
+## Handle exceptions
+
+We wouldn't want our program to crash when an exception is raised.
+
+Therefore, when we know a piece of code is error prone, we can wrap it in a `begin` block and rescue the exception with a `rescue` block.
+The `begin` block marks the beginning of the code that might raise an exception, and the `rescue` block handles the exception.
+
+```crystal
+begin
+ raise "This is an error"
+rescue
+ puts "An error occurred!"
+end
+```
+
+The `rescue` block can also be specified with a variable to get the exception object.
+
+```crystal
+begin
+ raise "This is an error"
+rescue ex
+ puts "An error occurred: #{ex.message}"
+end
+```
+
+The `rescue` block can also be specified with a specific exception type only to catch that exception.
+
+```crystal
+begin
+ raise ArgumentError.new("This is an argument error")
+rescue ArgumentError
+ puts "An argument error occurred!"
+end
+
+# or
+
+begin
+ raise ArgumentError.new("This is an argument error")
+rescue ex : ArgumentError
+ puts "An argument error occurred: #{ex.message}"
+end
+```
+
+Multiple `rescue` blocks can be used to handle different types of exceptions.
+In the example below, the first `rescue` block will catch an `ArgumentError`, and the second `rescue` block will catch any other exception.
+
+```crystal
+begin
+ raise ArgumentError.new("This is an argument error")
+rescue ArgumentError
+ puts "An argument error occurred!"
+rescue
+ puts "An error occurred!"
+end
+```
+
+The `begin` block can also have an `else` block, which is executed if no exception is raised.
+
+```crystal
+begin
+ puts "No error occurred"
+rescue
+ puts "An error occurred!"
+else
+ puts "No error occurred"
+end
+```
+
+Lastly, there is an `ensure` block that is always executed, regardless of whether an exception was raised.
+
+```crystal
+begin
+ raise "This is an error"
+rescue
+ puts "An error occurred!"
+ensure
+ puts "This is always executed"
+end
+```
+
+## Method convention
+
+If you have checked, specific methods have two versions, one with `!` and the other without.
+This can mean two different things.
+One is that the method mutates the object, and the other is that the method can raise an exception.
+
+But there is also another convention around ending a method with `?` mentioned in the previous concept.
+Some methods raise an exception by default but also have a version ending with `?` which returns `nil` instead of raising an exception.
+
+This is ideal when you want to avoid an error being raised. 
+This can benefit performance since it doesn't have to create a stack trace and, if setup correctly, could make the code safer.
```

Comment:
Out of curiosity, how would not using an error/exception improve safety?

Reply:
Part of that text is based of: https://crystal-lang.org/reference/1.13/syntax_and_semantics/exception_handling.html#alternative-ways-to-do-error-handling.

But using for example `nil` instead of throwing errors, so will the compiler make you handle `nil` meanwhile if you have code which throws error the compiler will not make you handle that instead if you are unlucky could make your application just halt execution.